js/packages/mobile-wallet-adapter-protocol-web3js/src/transact.ts

```diff
@@ -27,7 +27,7 @@ interface Web3SignTransactionsAPI {
 }
 
 interface Web3SignMessagesAPI {
-    signMessages(params: { address: Base64EncodedAddress; payloads: Uint8Array[] }): Promise<Uint8Array[]>;
+    signMessages(params: { addresses: Base64EncodedAddress[]; payloads: Uint8Array[] }): Promise<Uint8Array[]>;
```

Comment:
Now this has to return a list of signature arrays, right?

```typescript
Promise<Uint8Array[][]>;
```

Reply:
`Uint8Array[]`

TS syntax question: is this an array of uint8, or an array of uint8array?

Reply:
If it's the latter, then this returns an array of signed messages (where each message may have several signatures appended)

Reply:
`Array<Uint8Array>` and `Uint8Array[]` are synonyms. Those are arrays of `Uint8Array` instances.

Still, when we add multi-address signing, we need to end up in this scenario:

```
const foo = signMessages({addresses: [A,B,C], payloads: [Uint8Array([1]), Uint8Array([2]), Uint8Array([3])]);
console.log(foo);
```

```
[
  [signatureOfPayload1ByAddressA, signatureOfPayload1ByAddressB, signatureOfPayload1ByAddressC],
  [signatureOfPayload2ByAddressA, signatureOfPayload2ByAddressB, signatureOfPayload2ByAddressC],
  [signatureOfPayload3ByAddressA, signatureOfPayload3ByAddressB, signatureOfPayload3ByAddressC],
]
```

That's a `Uint8Array[][]` or synonymously an `Array<Array<Uint8Array>>`.

Reply:
This is also just an opinion. You could equally argue for:

```
[
  [signatureOfPayload1ByAddressA, signatureOfPayload2ByAddressA, signatureOfPayload3ByAddressA],
  [signatureOfPayload1ByAddressB, signatureOfPayload2ByAddressB, signatureOfPayload3ByAddressB],
  [signatureOfPayload1ByAddressC, signatureOfPayload2ByAddressC, signatureOfPayload3ByAddressC],
]
```

But still the type is `UintArray[][]`.

Reply:
Ah, I see now. This RPC method does not return signatures though; it returns the signed payload. This would be [`message1 || sigA || sigB || ... || sigN`, `message2 || sigA || sigB || ... || sigN`, ...]

Reply:
Alternative API: `Readonly<{[address: Base64EncodedAddress]: Uint8Array[]}>`

```
{
  A: [payload1SignatureByAddressA, payload2SignatureByAddressA, payload3SignatureByAddressA],
  B: [payload1SignatureByAddressB, payload2SignatureByAddressB, payload3SignatureByAddressB],
  C: [payload1SignatureByAddressC, payload2SignatureByAddressC, payload3SignatureByAddressC],
}
```

Reply:
We had spec'ed it out the current way to allow for the wallets and/or chain-specific extensions to specify how signatures are attached to messages. For eg, the message may need to be modified to add a domain prefix, wrapped in a PKCS#7 wrapper, etc.

Reply:
> This RPC method does not return signatures though; it returns the signed payload.

Ah, good point. This keeps the return types a bit more consistent with signed transactions.

Reply:
> This RPC method does not return signatures though; it returns the signed payload

Right right. Getting it mixed up with `@solana/wallet-adapter` which does return just the signature.

Reply:
Oof. OK, I remember having a conversation with one of you that went something like ‘does `signMessages` return `sign(sign(message, A), B)` or `[sign(message, A), sign(message, B)]` and I recall the answer was the latter.

¯\_(°ペ)_/¯

Reply:
`signMessage` and `signTransaction` work the same way -- all signers sign the same payload, and have their signatures added to the return. Each signature is independent.

Reply:
I think the problem with this conversation thread is that when you say ‘`signMessages` does…’ you have to specify whether you mean `wallet-adapter` or `mobile-wallet-adapter-protocol` because right now those two do different things.